uitools/common/src/Internal/PopupAttachmentImageProvider.cpp

```diff
@@ -0,0 +1,198 @@
+    emit finished();
+    return;
+  }
+
+  auto popupAttachment = item->popupAttachment();
+  if (popupAttachment->popupAttachmentType() != PopupAttachmentType::Image)
+  {
+    emit finished();
+    return;
+  }
+}
+
+PopupAttachmentImageResponse::~PopupAttachmentImageResponse() = default;
+
+QQuickTextureFactory* PopupAttachmentImageResponse::textureFactory() const
+{
+  QImage thumbnail;
+  QIcon icon;
+
+  if (m_popupAttachmentItem)
+  {
+    thumbnail = m_popupAttachmentItem->thumbnail();
+  }
+
+  // Handles edge case for Mobile since QAbstractFileIconProvider returns a Null QIcon.
+#if defined(Q_OS_IOS) || defined(Q_OS_ANDROID)
+  if (thumbnail.isNull() || m_popupAttachmentItem->popupAttachmentType() != PopupAttachmentType::Image)
+  {
+    switch (m_popupAttachmentItem->popupAttachmentType()) {
+      case PopupAttachmentType::Image:
+        icon = QIcon(":/esri.com/imports/Calcite/images/image.svg");
+        break;
+      case PopupAttachmentType::Video:
+        icon = QIcon(":/esri.com/imports/Calcite/images/video.svg");
+        break;
+      case PopupAttachmentType::Document:
+        icon = QIcon(":/esri.com/imports/Calcite/images/file.svg");
+        break;
+      case PopupAttachmentType::Other:
+        [[fallthrough]];
+      default:
+        icon = QIcon(":/esri.com/imports/Calcite/images/other.svg");
+        break;
+    }
+
+
+    auto pixmap = icon.pixmap(32, 32);
+    thumbnail = pixmap.toImage();
+    return QQuickTextureFactory::textureFactoryForImage(thumbnail);
+  }
+#endif
+
+  if (thumbnail.isNull() && m_popupAttachmentItem->dataFetched())
+  {
+    QAbstractFileIconProvider iconProvider;
+    QFileInfo fileInfo(m_popupAttachmentItem->localData().toLocalFile());
+
+    icon = iconProvider.icon(fileInfo);
+    auto pixmap = icon.pixmap(32, 32);
+    thumbnail = pixmap.toImage();
+  }
+
+  return QQuickTextureFactory::textureFactoryForImage(m_requestedSize.isValid() ? thumbnail.scaled(m_requestedSize) : thumbnail);
+}
+}
+
+const QString PopupAttachmentImageProvider::PROVIDER_ID = QStringLiteral("PopupAttachmentItems");
+
+PopupAttachmentImageProvider* PopupAttachmentImageProvider::instance()
+{
+  // It is assumed that the instance is passed to QQuickEngine image providers -- which will take
+  // ownership and delete this object, otherwise a memory leak will occur.
+  static PopupAttachmentImageProvider* self = new PopupAttachmentImageProvider;
+  return self;
+}
+
+PopupAttachmentImageProvider::PopupAttachmentImageProvider() :
+  m_internalObject(new QObject)
+{
+}
+
+PopupAttachmentImageProvider::~PopupAttachmentImageProvider()
+{
+  delete m_internalObject;
+}
+
+bool PopupAttachmentImageProvider::registerItem(PopupAttachmentItem* item)
+{
+  if (!item)
+    return false;
+
+  const auto key = item->id();
+
+  if (m_itemMap.contains(key))
+    return false;
+
+  m_itemMap.insert(key, item);
+  QObject::connect(item, &QObject::destroyed, m_internalObject, [this, key]
+  {
+    m_itemMap.remove(key);
```

Comment:
Is there any risk with the async image provider stuff that there could be any data races on `m_itemMap`? I assume the connections will either be direct, or queued if there is threading involved and everything will happen on the same thread. Thoughts?

Reply:
I don't believe so but I need to think about this a little more.